### PR TITLE
Reimplement loading indicator via IntersectionObserver

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial #30620.

TODO:
- [x] Tests

Key points:
1. Most refactoring is straightforward code move from `custom-element.js` to `loading-indicator.js`.
2. However, one critical change: the `inViewport` callback is no longer used and instead `IntersectionObserver` is used. Thus, the main API that custom element calls is: `track` (start tracking the element in case loading indicator needs to be shown) and `untrack` (stop tracking and cleanup).
3. The `isLoadingReused` is removed because it's not a property of an element but rather the property of a loader - whether or not the loader is shown for layout or for any other reason.
4. The `LoadingIndicator` service tracks the element's state in a weak map.
5. The `LoadingIndicator` service is installed _only_ in the environments where it's used. E.g. it's ignored for A4A or inabox.

### CODE REVIEW

File: extensions/amp-twitter/0.1/amp-twitter.js
Comment (samouri) [inline extensions/amp-twitter/0.1/amp-twitter.js:161]: great change

File: src/base-element.js
Comment (samouri) [inline src/base-element.js:704]: nit: I think the word we commonly use for this is `force`

File: src/custom-element.js
Comment (samouri) [inline src/custom-element.js:1621]: nit: don't we usually allow the passing of an `Element` here? E.g. `this.element_`

File: src/service.js
Comment (samouri) [inline src/service.js:70]: can you explain this rename? I don't have much context on what internal means. Does it mean a non-amp element (node within an amp elem). I'm guessing the addition of `AmpDoc` as a potential param is an optimization?

File: src/service/core-services.js
Comment (samouri) [inline src/service/core-services.js:146]: nit: can you either give the _why_ or remove the comment. as-is it doesn't tell you anything beyond what is obvious in the code :)

File: src/service/loading-indicator.js
Comment (samouri) [inline src/service/loading-indicator.js:57]: The current `viewportCallback` is based on a loading rect with a 25% expanded rect. My gut says the safest way to proceed for all of our replacements is to:
1. First recreate as close to the same behavior as possible. In this case it means adding a 25% margin.
2. After landing, if we want, then we can turn up an experiment where we remove the margin and ensure it doesn't negatively affect metrics.

https://github.com/ampproject/amphtml/blob/6faff93057d642e9ecd977faddda85c033d72bf7/src/service/resources-impl.js#L1335-L1340

File: src/service/loading-indicator.js
Comment (samouri) [inline src/service/loading-indicator.js:123]: super nit: from the name alone I was surprised to see that it also creates dom nodes. the overall structure of the code still feels right to me though!

**Review by samouri — COMMENTED** [review]
Looks great so far

File: src/custom-element.js
Comment (dvoytenko) [reply src/custom-element.js:1621]: Either or is ok. The `ampdoc` is faster, b/c when an element is provided, it's first resolved to the `ampdoc` and then the service is looked up. So we are reusing the `ampdoc` lookup here done before.

File: src/service.js
Comment (dvoytenko) [reply src/service.js:70]: I initially renamed it to `internal` b/c it was not used outside of this file. But now it is. Also, per `TODO` there, the `getExistingServiceForDocInEmbedScope` will be replaced with this call everywhere. It's done in the https://github.com/ampproject/amphtml/issues/22733. So there's a slight race between two things I'm working on :)

File: src/service/loading-indicator.js
Comment (dvoytenko) [reply src/service/loading-indicator.js:57]: I think you're right here. It was initially expanded given manual measures. But it's probably better to make this consistent for now.

File: src/service/loading-indicator.js
Comment (dvoytenko) [reply src/service/loading-indicator.js:124]: I don't think so. InOb is pretty good at firing when the page is not super busy.

File: src/service/loading-indicator.js
Comment (dvoytenko) [reply src/service/loading-indicator.js:123]: Renamed it to `createLoaderState_`. Does this help?

**Review by samouri — COMMENTED** [review]
Code LGTM! Would still like to manually test before merge (and have unit tests) before merge.

**Comment by dvoytenko** [comment]
@samouri tests have been completed.

File: extensions/amp-animation/0.1/test/test-amp-animation.js
Comment (samouri) [inline extensions/amp-animation/0.1/test/test-amp-animation.js:38]: Does `??` get transformed properly? I'm not sure which new syntax we can / can't use yet.

Also, more important: why does `test-amp-animation` need any changes at all?

File: test/unit/test-amp-inabox.js
Comment (samouri) [inline test/unit/test-amp-inabox.js:43]: This seems fishy to me. I don't have the background here, but do you know why we are adding all the services by regular amp and then removing vs. only installing the correct ones to begin with?

File: test/unit/test-custom-element.js
Comment (samouri) [inline test/unit/test-custom-element.js:1928]: why is `toggle.resetHistory` necessary?

File: test/unit/test-loading-indicator.js
Comment (samouri) [inline test/unit/test-loading-indicator.js:147]: FMI: Is there a performance cost to leaving loaders on out-of-viewport items?

File: src/service/loading-indicator.js
Comment (samouri) [reply src/service/loading-indicator.js:142]: In that case we can remove the `startTime` concept - I don't think it serves any purpose if we no longer have a delay

**Review by samouri — APPROVED** [review]
Just some nits and questions. I manually tested the indicator on an `amp-list` examples page, including refresh, and LGTM

File: src/service/loading-indicator.js
Comment (samouri) [inline src/service/loading-indicator.js:63]: What do you think is a reasonable plan for improving this situation? When the polyfill supports rootMargin in x-origni iframes, do you think its reasonable to use that instead of the 3 situations we have today

File: src/service/loading-indicator.js
Comment (dvoytenko) [reply src/service/loading-indicator.js:142]: Yeah. I just wanted to avoid refactoring spilling deeper into the insides of the actual implementation of the loader. Are you ok I leave it up to some future rewrite?

File: extensions/amp-animation/0.1/test/test-amp-animation.js
Comment (dvoytenko) [reply extensions/amp-animation/0.1/test/test-amp-animation.js:38]: Yes. The `??` is fully available for us now. As for the `test-amp-animation` changes - it's tricky. This test has stubbed out the full implementation of the `IntersectionObserver`, but it was passing really weird data.

File: test/unit/test-amp-inabox.js
Comment (dvoytenko) [reply test/unit/test-amp-inabox.js:43]: We not removing services. This test first collects all regular AMP services and then inabox services and then compares them all. There should be some meaningful reason why one environment should miss some of the services. But loading indicator is a legitimate case here. So I'm basically testing that `ampSevices - loadingIndicator == inaboxServices`.

File: test/unit/test-custom-element.js
Comment (dvoytenko) [reply test/unit/test-custom-element.js:1928]: Some the tests above test specifically connection/build loading indicator. But here I wanted to exclusively test only the loading chain and how it works with the loading indicator. This is actually an old test. But the old test only checked that the loading indicator is eventually reset. But I wanted to test a bit more here: that it's first set and then reset. So instead of only testing the last value, I'm testing all toggles during the loading process.

File: test/unit/test-loading-indicator.js
Comment (dvoytenko) [reply test/unit/test-loading-indicator.js:147]: Yes. They continue to animate. Eventually I hope that some of it will be manageable using `content-visibility`. But for now this has to be done manually. And it's the old behavior as well.

File: src/service/loading-indicator.js
Comment (dvoytenko) [reply src/service/loading-indicator.js:63]: I think we need to delegate this decision to the full loader "intersect-resources" launch/cleanup. If we lose 25% margin for inViewport in some cases is a lot less critical vs losing loader functionality. So, in that scope we will have to decide what we do. It's very possible that we may chose to:
1. Add `root=document` feature to the polyfill. This would be a relatively straightforward thing to do. It's actually mostly done already.
2. Override of the native InOb that doesn't support `root=document` with the polyfill. This is, of course, a lot more controversial. For now this is a majority case, though I think it should be reducing quickly. But, certainly, using the polyfill where the native is available is super inefficient.

To remedy this, we could also be very selective when we override. E.g. we can override _only_ when we are iframed and _only_ when `root=document` is requested. E.g.

```
class IntersectionObserverSelector {
  constructor(callback, options) {
    const constr =
      isIframed() && options.root ?
      IntersectionObserverPolyfill :
      IntersectionObserverNative;
    return new constr(callback, options);
  }
}
```

But, this kind of effort is only worth it for loader, not inViewport.

**Pull request merged**
